File: algo/lc.121.cc
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.empty()) return 0;
        
        /*
        int maxdiff = 0;
        for (int i = 0; i < prices.size(); i++) {
            for (int j = i+1; j < prices.size(); j++) {
                maxdiff = max(maxdiff, prices[j] - prices[i]);
            }
        }
        return maxdiff;
        */
        // Time out O(n^2)
        
        int minprice = INT_MAX;
        int maxprofit = 0;
        for (auto i : prices) {
            minprice = min(minprice, i);
            maxprofit = max(maxprofit, i - minprice);
        }
        return maxprofit;
    }
};

File: algo/lc.121.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) < 2:
            return 0
        # at least two numbers
        maxprofit = 0
        lowp = max(prices)
        for x in prices:
            if x < lowp:
                lowp = x
            if x - lowp > maxprofit:
                maxprofit = x - lowp
        return maxprofit
